File: mcp-server/src/queries.ts
import { getDb, type ProjectRow, type SiteRow, type AuditRow, type SnapshotRow, type CredentialRow } from './db.js'

export function listProjects(): ProjectRow[] {
  return getDb().prepare(`
    SELECT p.id, p.name, p.description, p.color, p.icon, p.is_active,
           COUNT(s.id) AS site_count, p.created_at
    FROM projects p
    LEFT JOIN sites s ON s.project_id = p.id
    GROUP BY p.id
    ORDER BY p.created_at DESC
  `).all() as ProjectRow[]
}

export function listSites(projectId?: number): SiteRow[] {
  const base = `
    SELECT s.id, s.project_id, p.name AS project_name,
           s.name, s.domain, s.url,
           s.gsc_property, s.ga4_property, s.bwt_property, s.clarity_project,
           s.notes, s.is_active,
           da.score AS last_audit_score, da.crawled_at AS last_audit_at,
           s.created_at
    FROM sites s
    JOIN projects p ON p.id = s.project_id
    LEFT JOIN (
      SELECT site_id, score, crawled_at
      FROM dom_audits
      WHERE id IN (SELECT MAX(id) FROM dom_audits GROUP BY site_id)
    ) da ON da.site_id = s.id
  `
  if (projectId !== undefined) {
    return getDb().prepare(base + ' WHERE s.project_id = ? ORDER BY s.created_at DESC')
      .all(projectId) as SiteRow[]
  }
  return getDb().prepare(base + ' ORDER BY s.created_at DESC').all() as SiteRow[]
}

export function getSiteByDomain(domain: string): SiteRow | null {
  return getDb().prepare(`
    SELECT s.id, s.project_id, p.name AS project_name,
           s.name, s.domain, s.url,
           s.gsc_property, s.ga4_property, s.bwt_property, s.clarity_project,
           s.notes, s.is_active,
           da.score AS last_audit_score, da.crawled_at AS last_audit_at,
           s.created_at
    FROM sites s
    JOIN projects p ON p.id = s.project_id
    LEFT JOIN (
      SELECT site_id, score, crawled_at
      FROM dom_audits
      WHERE id IN (SELECT MAX(id) FROM dom_audits GROUP BY site_id)
    ) da ON da.site_id = s.id
    WHERE s.domain = ? OR s.domain = ?
  `).get(domain, domain.replace(/^www\./, '')) as SiteRow | null
}

export function getLastAudit(siteId: number): AuditRow | null {
  return getDb().prepare(`
    SELECT * FROM dom_audits
    WHERE site_id = ?
    ORDER BY id DESC LIMIT 1
  `).get(siteId) as AuditRow | null
}

export function getAuditHistory(siteId: number, limit = 10): AuditRow[] {
  return getDb().prepare(`
    SELECT * FROM dom_audits
    WHERE site_id = ?
    ORDER BY id DESC LIMIT ?
  `).all(siteId, limit) as AuditRow[]
}

export function getSnapshot(siteId: number, source: string): SnapshotRow | null {
  return getDb().prepare(`
    SELECT * FROM mcp_snapshots
    WHERE site_id = ? AND source = ?
    ORDER BY fetched_at DESC LIMIT 1
  `).get(siteId, source) as SnapshotRow | null
}

export function getAllSnapshots(siteId: number): SnapshotRow[] {
  return getDb().prepare(`
    SELECT * FROM mcp_snapshots
    WHERE site_id = ?
    ORDER BY source, fetched_at DESC
  `).all(siteId) as SnapshotRow[]
}

export function getCredentials(): CredentialRow[] {
  return getDb().prepare(`
    SELECT service, label, credential_type, is_active, last_tested, last_error
    FROM credentials
  `).all() as CredentialRow[]
}

export function searchSites(query: string): SiteRow[] {
  const q = `%${query}%`
  return getDb().prepare(`
    SELECT s.id, s.project_id, p.name AS project_name,
           s.name, s.domain, s.url,
           s.gsc_property, s.ga4_property, s.bwt_property, s.clarity_project,
           s.notes, s.is_active,
           da.score AS last_audit_score, da.crawled_at AS last_audit_at,
           s.created_at
    FROM sites s
    JOIN projects p ON p.id = s.project_id
    LEFT JOIN (
      SELECT site_id, score, crawled_at
      FROM dom_audits
      WHERE id IN (SELECT MAX(id) FROM dom_audits GROUP BY site_id)
    ) da ON da.site_id = s.id
    WHERE s.domain LIKE ? OR s.name LIKE ? OR p.name LIKE ?
    ORDER BY s.domain
  `).all(q, q, q) as SiteRow[]
}

File: src-tauri/src/db/queries.rs
// SQL query helpers — thin wrappers over db::with_conn used by commands.
use anyhow::Result;
use crate::db::with_conn;
use crate::models::{project::Project, site::Site};

// ── Projects ─────────────────────────────────────────────────────────────────

pub fn list_projects() -> Result<Vec<Project>> {
    with_conn(|conn| {
        let mut stmt = conn.prepare(
            "SELECT p.id, p.name, p.description, p.color, p.icon, p.is_active,
                    COUNT(s.id) AS site_count,
                    p.created_at, p.updated_at
             FROM projects p
             LEFT JOIN sites s ON s.project_id = p.id
             GROUP BY p.id
             ORDER BY p.created_at DESC",
        )?;
        let rows = stmt.query_map([], |r| {
            Ok(Project {
                id: r.get(0)?,
                name: r.get(1)?,
                description: r.get(2)?,
                color: r.get(3)?,
                icon: r.get(4)?,
                is_active: r.get(5)?,
                site_count: r.get(6)?,
                created_at: r.get(7)?,
                updated_at: r.get(8)?,
            })
        })?;
        rows.collect::<rusqlite::Result<Vec<_>>>().map_err(Into::into)
    })
}

// ── Sites ─────────────────────────────────────────────────────────────────────

pub fn list_sites(project_id: Option<i64>) -> Result<Vec<Site>> {
    with_conn(|conn| {
        let sql = if project_id.is_some() {
            "SELECT s.id, s.project_id, p.name AS project_name,
                    s.name, s.domain, s.url,
                    s.gsc_property, s.ga4_property, s.bwt_property, s.clarity_project,
                    s.notes, s.is_active,
                    da.score AS last_audit_score, da.crawled_at AS last_audit_at,
                    s.created_at, s.updated_at
             FROM sites s
             JOIN projects p ON p.id = s.project_id
             LEFT JOIN (
                 SELECT site_id, score, crawled_at
                 FROM dom_audits
                 WHERE id IN (SELECT MAX(id) FROM dom_audits GROUP BY site_id)
             ) da ON da.site_id = s.id
             WHERE s.project_id = ?1
             ORDER BY s.created_at DESC"
        } else {
            "SELECT s.id, s.project_id, p.name AS project_name,
                    s.name, s.domain, s.url,
                    s.gsc_property, s.ga4_property, s.bwt_property, s.clarity_project,
                    s.notes, s.is_active,
                    da.score AS last_audit_score, da.crawled_at AS last_audit_at,
                    s.created_at, s.updated_at
             FROM sites s
             JOIN projects p ON p.id = s.project_id
             LEFT JOIN (
                 SELECT site_id, score, crawled_at
                 FROM dom_audits
                 WHERE id IN (SELECT MAX(id) FROM dom_audits GROUP BY site_id)
             ) da ON da.site_id = s.id
             ORDER BY s.created_at DESC"
        };

        let mut stmt = conn.prepare(sql)?;
        let rows = if let Some(pid) = project_id {
            stmt.query_map(rusqlite::params![pid], map_site)?
        } else {
            stmt.query_map([], map_site)?
        };
        rows.collect::<rusqlite::Result<Vec<_>>>().map_err(Into::into)
    })
}

fn map_site(r: &rusqlite::Row) -> rusqlite::Result<Site> {
    Ok(Site {
        id: r.get(0)?,
        project_id: r.get(1)?,
        project_name: r.get(2)?,
        name: r.get(3)?,
        domain: r.get(4)?,
        url: r.get(5)?,
        gsc_property: r.get(6)?,
        ga4_property: r.get(7)?,
        bwt_property: r.get(8)?,
        clarity_project: r.get(9)?,
        notes: r.get(10)?,
        is_active: r.get(11)?,
        last_audit_score: r.get(12)?,
        last_audit_at: r.get(13)?,
        created_at: r.get(14)?,
        updated_at: r.get(15)?,
    })
}
